### ISSUE TITLE
style: decide on whitespace rules for . , ; ( ) [ ] { }

### ISSUE THREAD

**Issue by snappizz** [supercollider/supercollider#2931]
let's decide on the right way to whitespace the characters `. , ; ( ) [ ] { }`. here are my proposals. might as well lump them all together in one issue because life is short

## semicolons

never put whitespace immediately before `;`.

```supercollider
x = 3 + 5 ; // incorrect
x = 3 + 5;  // correct
```

## commas

never put whitespace immediately before a comma. unless the comma is at the end of a line, always follow it with a single space.

```supercollider
// correct:
f.value(3, 4);
f.value(
    3,
    4
);

// incorrect:
f.value(3 , 4);
f.value(3,4);
```

## dots in method calls

inline, don't put any whitespace around `.`

```semicolon
// correct
f.value(3, 4);
{ SinOsc.ar }.play;

// incorrect
f . value(3, 4);
```

you can also use a multiline form for `.` in the idiomatic case of chaining method calls on the same object. the dot should be on the second line, not the first, and it should be indented one level more than the indentation level of the receiver. there should be no whitespace after the dot.

```semicolon
Button()
    .states_([["blorp", nil, nil]])
    .action_({
        "hey hey hey".postln
    });
```

## brackets, inline

all of these three bracket types can be used in inline or multiline form. in inline form, don't put any whitespace immediately inside `( ) [ ]`, but do put one space after `{` and one space before `}`. since `{ }` delimit functions, it's "heavier" than the other two.

```supercollider
// correct:
a = f.value(10);
b = [1, 2, 3];
c = b.collect({ |x| x + 3 });

// incorrect:
a = f.value( 10 );
b = [ 1, 2, 3 ];
c = b.collect({|x| x + 3});
```

in some cases you need an empty `{ }`. in that case, use one space inside.

```supercollider
MyBuffer : Buffer { }
```

## brackets, multiline

K&R for all three.

```supercollider
a = x + f.value(
    some_complicated_expression
);
b = [
    "eggs",
    "milk",
    "bread",
];
c = {
    var x;
    x = y + 1;
};
```

## more method calls

i'll be brief for these ones. we won't get into binary operators in this issue.

```supercollider
// don't put whitespace before ( )
foo(bar);
bar.foo();
Mix();
someFunction.();

// buffer curly-brace functions with exactly one space
// (but again, don't put space before a semicolon)
if(this) { that } { dang };
Routine.run { 3.wait; };
```

**Comment by jamshark70** [comment]
> you can also use a multiline form for . in the idiomatic case of chaining method calls on the same object. the dot should be on the second line, not the first, and it should be indented one level more than the indentation level of the receiver. there should be no whitespace after the dot.

I agree with this, but this will require a change to SC-IDE's auto-indenter. Currently, pressing tab with a code block selected will clobber all of your manually carefully indented method chains.

Currently we can't consider that truly a bug... but if we decide on code formatting standards and the IDE's behavior conflicts, then it would become a bug.

**Comment by vivid-synth** [comment]
What would you think about calling this a "style guide" instead of "the right way"? As much as things like ```f . value(3, 4);``` make my eyes hurt a little, I'm kinda concerned about being a little too overbearing with rules.

I think we should strive to be a fun environment to submit code to, and if someone submits high-quality code with some weird spaces before semicolons or something, I'd like to merge it (or I guess offer to fix it myself).

**Comment by snappizz** [comment]
[it's already called that](https://github.com/supercollider/supercollider/wiki/%5BWIP%5D-Code-style-guidelines)

**Comment by brianlheim** [comment]
I think he meant in your original post, @snappizz

> let's decide on the right way to whitespace the characters

And I agree with @vivid-synth; especially when it comes to whitespace I think we should be more on the forgiving side and decide on a few things (if any) we want to lint. I'm not sure what that means for the interpretation of "rule" and "recommendation" in the guidelines; @vivid-synth maybe you have some thoughts on that?

Anyway, I think the original suggestions are all deserving of rule status except the guidelines around spaces in brackets, which I think should be recommendations at best. Lately I've been feeling that `a.value( x )` is much more comfortable than `a.value(x)`, and I wouldn't argue with anyone who wants to type the extra spaces. I feel even stronger about it when it comes to nested elements:

```supercollider
[ [1, 2], [3, 4] ];
a.value( x*3 );
b.value( (a.asString + "to") ++ (b.isNil) );
[ (x+1) * 2, 3 * (4+y) ]
```

I think that, in terms of what I'd like to see in a guideline, these would be at least acceptable, if not preferred. OTOH, sometimes I actually don't need the spaces in an integer array:
`[1,2,3,4,5]` and `[1, 2, 3, 4, 5]` are both okay with me. I still agree that we should prefer with spaces, but I'm trying to express the range of what I consider acceptable here.

**Comment by jamshark70** [comment]
> Lately I've been feeling that `a.value( x )` is much more comfortable than `a.value(x)`

Well, I can't bring myself to agree with that ("much?") but I wouldn't think my own preference (for the latter) is necessarily worth enforcing as a rule for every user running a linter on her own code.

That's a different question from standards for the main class library, where I would say, either we have rules or we don't. We shouldn't have user A contribute line 237 with `x.run(false)` and then user B follow up in line 238 with `x.set( \y, 1 )`. If we have to make a decision, I'd suggest against changing the 15+ years of existing practice in the classlib merely because one developer only "lately" comes to feel extra spaces are better. (Thrill of novelty perhaps.) And FWIW, I find the extra spaces for parens and brackets to be immensely distracting (and I've already made the point that spaces inside braces help distinguish braces from parens at smaller font sizes, which benefit we lose if parens acquire spaces as well).

**Comment by brianlheim** [comment]
> ("much?")

Yes, "much."

> I'd suggest against changing the 15+ years of existing practice in the classlib merely because one developer only "lately" comes to feel extra spaces are better.

Agreed, I hope I didn't make it sound like I was suggesting this?

> at smaller font sizes, which benefit we lose if parens acquire spaces as well

or just increase the font size... :P

**Comment by jamshark70** [comment]
It's really fascinating in this discussion how people have very different opinions about this.

> or just increase the font size...

No, I don't think so. There is benefit in seeing larger blocks of code at one time.

Outside the Mac-osphere, laptops' built-in screens are commonly 768 pixels high. Counting pixels in GIMP, I'm getting 14px per line = 768/14 = a maximum of almost 55 lines on screen at once. I reserve part of the screen width for GUIs, so I put the post window at the bottom, leaving about 25 lines visible with my current setup (after the XFCE app-menu bar, IDE window title bar, menu bar, document tabs, post window header bar, and IDE status bar chew up some of the vertical space). I don't think I should have to reduce that further, or put the post window to the side and put my GUIs underneath the IDE window, because of a suggestion that one should use a larger font to compensate for spacing choices that impair readability. (If the next bit of advice were to be "get a Mac, then," that would seem an exercise in *reductio ad absurdum* :rofl: )

![spacing-samples](https://user-images.githubusercontent.com/318301/26998956-4f2e48e8-4dc0-11e7-8ac5-f9a34c4a4446.png)

```
a = { |extent| Rect.fromPoints(Point(0, 0), extent) };

a = { | extent | Rect.fromPoints( Point( 0, 0 ), extent ) };
```

No. 2.

- *What* is that free-floating closing-paren doing? Serious question, actually. Why is this a good thing?
- What, visually, discourages reading `extent | Rect` vs `| extent |`?

```
a = {|extent| Rect.fromPoints(Point(0, 0), extent)};
```

No. 3. There's old code floating around with things like `n.do({ |i| ... })`, and it was always easy to distinguish `})` from `) }` -- but by removing spaces, it becomes harder to distinguish them. Maybe it's not as bad when adding more spaces, but it's a similar problem.

```
a = {|extent|Rect.fromPoints(Point(0,0),extent)};
```

No. 4. This is basically unreadable -- but, incredibly, I've seen code posted to the mailing list that looks exactly like this.

I think we all agree nos. 3 and 4 should be discouraged (and it's obvious why, in the screenshot). No. 2 bothers me. I think I've provided a concrete reason why I don't like it (distinguishing glyphs that have a number of surface features in common). The defense of more-spaces has been that it "seems more comfortable" -- which, I apologize if I'm being dense but I don't know quite what that means.

**Comment by jamshark70** [comment]
To be fair, I guess no. 2 could eventually grow on me... but I think I would still resist choosing a larger or different font only for `()` vs `{}` (and, no time this morning for a survey of parens and braces in programmers' fonts), when we already have spacing practices in the class library that help.

**Comment by telephon** [comment]
having thought about it a little, I think  (personally) I prefer no spaces around the arg names, but a space after `{`.

`a = {  |extent| Rect.fromPoints(Point(0, 0), extent) };`

I think I'd always use that if I have the choice. But well …

**Comment by snappizz** [comment]
can i get confirmation that the rules for semicolons, periods, and the use of K&R style for multiline brackets are uncontroversial?

**Comment by brianlheim** [comment]
++ fine with me

re spaces in parens: "more comfortable" means it's easier to separate tokens from one another. i'm used to reading supercollider code in this less spacey format anyway (although it still makes me feel claustrophobic), so it wouldn't bother me to keep it the way you and Julian and Nathan prefer it.

**Comment by jamshark70** [comment]
> can i get confirmation that the rules for semicolons and periods and the use of K&R style for multiline brackets are uncontroversial?

Complete agreement from me.

> re spaces in parens: "more comfortable" means it's easier to separate tokens from one another.

Hm, I wonder if that's a font-choice issue? Most fonts for publishing have at least a slight descender on parens, and sometimes the top of the parens rises above the top of capital letters too -- so I catch the separation by the deviation from the letters' baseline. I'm pretty sure I've seen some programmers' fonts where parens' bottom and sits exactly on the baseline. In that case, you might need the extra horizontal space.

In SCIDE, I'm using Inconsolata, which *is* meant for publishing and it definitely has a slight descender.

**Comment by brianlheim** [comment]
Let it go, please.

**Comment by smrg-lm** [comment]
> - What, visually, discourages reading `extent | Rect` vs `| extent |`?

I think that gestalt issue is they key concept.

>`b.value( (a.asString + "to") ++ b.isNil );`
`[ (x+1) * 2, 3 * (4+y) ]`

here spaces are in the outer parenthesis only helping grouping compound expressions. I see this distribution as "better" because grouping is coherent with the programming structures (even the lack of spaces between internal operators)

`[(x+1) * 2, 3 * (4+y)]`

this is somehow unbalances because inner spaces.

> `a.value( x )`
`x.set( \y, 1 )`

these would not need spaces, grouping is clear because symbols are different enough and expressions are simple. However:

>`[1,2,3,4,5]`

this goes against any ethical and moral norm, "you shall use spaces after commas in lists", that is out of discussion.

**Comment by smrg-lm** [comment]
Another thing, syntax shortcuts are also an style issue:

```
a = { arg x; };
a = { | x | };

10.do { };
10.do({ });
```
and control flow structures. When teaching I try to skip any syntax shortcut so programming constructs are clear and consistent (`arg` is consistent with `var`, `classvar` and `const`, parenthesis show that the function is a parameter), but how to use it in the class library is kind of important too. I would not use shortcuts unless they are more computationally efficient somehow...

**Comment by telephon** [comment]
I see your point. In general, spaces even it out, and a bit of care:

instead of:
>`[ (x+1) * 2, 3 * (4+y) ]`

`[(x + 1) * 2, (4 + y) * 3]`

**Comment by smrg-lm** [comment]
Yes, that balances in a different way, at least for me. James's point is the key point for me although. It has to do with the shape of the symbols and it's spacing arrangement regarding it's meaning and most common appearance combinations. However there will be always ambiguous cases, different perceptions and taste involved. My last sentence was a joke but shows that the symbols's shape are significantly differents to have no confusion but... if the array was of floating point numbers with changing precision: points are similar to commas and grouping of digits would be less clear. The only non controvertial problem would be inconsistency, i.e. applying different conventions for the same cases, but to define what are different cases seems to be difficult.

**Comment by telephon** [comment]
yes, that is true, it all depends on context.

**Comment by brianlheim** [comment]
Here are some statistics on the class library. Numbers are in LOC matching the pattern.

### spaces after `{`

no space: 19k
**space: 140k**

### spaces after `(`

**no space: 210k**
space: 19k

### spaces after `[`

**no space: 36k**
space: 1.5k

### space between chars in `({`

**no space: 10k**
space: <1k

### space between chars in `{(`

no space: 109
**space: 772**

### space between chars in `{|`

no space: 9k
**space: 30k**

### space between chars in `[(`

no space: 45
space: 15

--- 

At this point, I personally would be happy to take the more frequent usages to be the guidelines for now. Rereading this thread it seems like consensus is that, modulo nesting and corner cases, they are preferable. Specific cases may come up where we want to revisit this, but my opinion right now it would be better to spend time on other points of style. As you can see from these numbers, some of these cases are quite rare. Is that OK with everyone?

The space around operators topic would be a good one to visit next.

**Comment by prko** [comment]
Hi,
the post window of SC-IDE posts the parentheses of arrays and the parentheses of dictionaries with whitespaces.
```supercollider
[1, 2, 3]
```
returns
`-> [ 1, 2, 3 ]`
```supercollider
(test: 1) 
```
returns
`-> ( 'test': 1 )`
I think it is a bit strange that the recommended code style and the post window do not have a unified style.
What do you think about it?

**Comment by snappizz** [comment]
@prko i wish we could, but i think it's too dangerous to suddenly change that and break backward compatibility. who knows what code could be dependent on this behavior. we recently changed `Float:asString` to use a decimal point and ended up breaking the ambisonic toolkit.

@brianlheim sounds like a deal. i think we have reached consensus here, so i'm closing.

**Comment by brianlheim** [comment]
> @brianlheim sounds like a deal. i think we have reached consensus here, so i'm closing.

Thanks! I wish I had looked at this sooner, I didn't realize my comment had gotten a thumbs-up consensus. :)

**Comment by tttptiii** [comment]
I know this is an old issue, but I'm leaving a comment here since I think this question is related to the whitespace rules:

Is there a style guide for key-value pairs, like assigning values for kwargs in functions(e.g. `f.value(b: 2, a: 10);`)?

I found inconsistent use of whitespace after colon in [this document](https://github.com/supercollider/supercollider/blob/develop/HelpSource/Tutorials/Getting-Started/04-Functions-and-Other-Functionality.schelp) (123rd line, `f.value(2, c:3, b:4, d: 1); // (2 + 4) * 3 - 1`), so I tried to format the document according to the [Style Guidelines: SuperCollider · supercollider/supercollider Wiki](https://github.com/supercollider/supercollider/wiki/Style-Guidelines%3A-SuperCollider), but I couldn't find the guidline related to such formatting in the document.
